### ISSUE TITLE
Define new process for adding dependencies

### ISSUE THREAD

**Issue by nmittler** [istio/istio#5023]
Now that vendors has been merged into istio/istio (no longer a submodule), let's use this issue as a forum for discussion on what the new process for adding deps should be.

**Comment by douglas-reid** [comment]
Moving forward, I suggest that we determine the process for the new way _before_ switching to the new way.

I'm not sure what the rush was, but it seems like we could have at least had a proposed / trial process in place before jumping.

**Comment by ldemailly** [comment]
I would start with rather urgently emailing istio-dev@ about the change. Even if you haven't figured out all the details or to say as much.

**Comment by ldemailly** [comment]
I closed the PRs in vendor-istio and archived that repo. It was a fun/learning for 2 months but I'm glad new people take on the (ungrateful) work.

ps: we could have dropped 2 steps in the multi step dance using merge commit instead of squash commit; I successfully did that with a PR yesterday

**Comment by nmittler** [comment]
@douglas-reid agreed ... my intent wasn't to have the PR merged yet, but to iron out the kinks first.  But it is what it is at this point.

**Comment by sbezverk** [comment]
@nmittler I have a need to add a couple of libraries which were missing from k8s client-go, whitout them I cannot compile.  Should I run `dep ensure -add k8s.io/client-go/informers` and then commit vendor changes to my pr? Please advise..

**Comment by nmittler** [comment]
@sbezverk yes ... you can send the PR to me.

**Comment by nmittler** [comment]
@rshriram @geeknoid just to start off the discussion....

From [this thread](https://github.com/istio/istio/pull/5014#issuecomment-382109418) it sounds like we still want 2 PRs ... is that right?

Why do we think dependency changes need to be reviewed separately?

**Comment by sbezverk** [comment]
I think it should be done in the same PR as the code which is requering these libraries. If it is done in a separate PR, `dep` will not find references in the code to those libraries and will give a warning message that with the next update, "unused" libraries  will be removed.

**Comment by rshriram** [comment]
Dep is not running in ci/cd anyway. So I don’t think it matters ?

The reason for separate PRs is because of github UI. 
dependency update PRs can simply be approved and merged without much code review. Even if it’s 20K lines. But if there is some 3-line code change in istio that is using a 20K loc dependency, having both code and dependency changes in same PR makes it very hard to review the istio code change.

Alternatively, we could do everything in one PR as long as the author first submits code changes only, gets approval and then adds a commit with dependency changes, gets lgtm/merge.

**Comment by costinm** [comment]
How about: case by case. 

The main reason for having a separate PR is if we have a different approval process - and we certainly should have this for any new dependency. Projects like Apache requires any new dependency to be 
approved - with a discussion on the list, and clear rules on licence, maturity, who maintains it, etc.

If someone adds few files from an existing dependency - usually result of pruned packages - and the reviewer is ok with looking at the modified files ( and it's not 20k lines ) - why make things complicated ? 
Updating a dependency to a newer version should also be ok. 

Even with vendor in submodule a reviwer could still see the changes in the deps - together with the PR, and I think it makes it easier to understand both the PR and the vendor change if they are together.

My only concern is with raising the bar on new dependencies.

**Comment by rshriram** [comment]
Fair point. May be we see how things go for the next two weeks. Gather data and establish the process. 

With regard to vendor.check, doesn’t that beat the purpose of checking in vendor? If the code compiles and passes all tests, why bother checking against dep output in the first place ? Am I missing something here?

**Comment by douglas-reid** [comment]
I guess I'm not clear on what a review of vendor code is aiming to accomplish entirely. I was hoping there was a process that is verifying the dependency code being committed is *actually* the code from the repo, and not modified in any way/shape/form.

Separate PRs for adding dependencies and using dep ensure is a recipe for pain -- exactly as described in #5014.

> The time it takes between the submission of the 2 PRs presents a general problem to the developer workflow. 

If the change to remove vendor/ submodule was meant to address the separate PRs problem, it seems to me that we should not now encourage that very pattern with our new approach.

**Comment by nmittler** [comment]
> I was hoping there was a process that is verifying the dependency code being committed is actually the code from the repo, and not modified in any way/shape/form.

Agreed. Just to be clear, are you suggesting the re-introduction of `vendor.check` for this purpose? And to have this run by CI?

**Comment by douglas-reid** [comment]
@nmittler Sure, if that works. My main goal is to add a little peace of mind while reducing the onus on the reviewer. It _feels_ like there should be an automated way to validate that the dependency was added correctly and completely. And a little presubmit that runs on vendor-change PRs to give reviewers the "the proposed dependency is valid and has not been changed from source" flag sounds good to me. 

I'm open to other suggestions as well.

**Comment by nmittler** [comment]
@douglas-reid I've created a PR to add a diff check on the vendor files: https://github.com/istio/istio/pull/5051

**Comment by nmittler** [comment]
Just had a chat with @costinm and he's expressed concerns of running `dep ensure` on every PR.  He'd prefer to run it nightly and on a pre-release.

@douglas-reid @rshriram Any thoughts?

**Comment by nmittler** [comment]
@rshriram is there a way in CircleCI to detect whether a given change affects the `vendor` directory and then optionally execute a rule based on that?

**Comment by costinm** [comment]
Left few comments in the PR, it seems reasonable to do it, but part of
lint, not as a dep or before other jobs.

Nate found that dep update check ( in periodic) has been broken, no longer
checking that upgrading the deps is possible. Need to fix that as well.

**Comment by nmittler** [comment]
@ldemailly where is the appropriate place to add/update documentation for the new process?

**Comment by ldemailly** [comment]
Same as the previous place: wiki Dev guide and Vendor faq

**Comment by nmittler** [comment]
Vendor FAQ updated: https://github.com/istio/istio/wiki/Vendor-FAQ

**Comment by nmittler** [comment]
And Dev guide changes: https://github.com/istio/istio/wiki/Dev-Guide

@ldemailly is that everything?

**Comment by ldemailly** [comment]
sgtm, thx
